From 4552a1d6c8c1335e94c7bdfb1d6846249f706743 Mon Sep 17 00:00:00 2001
From: Sylvain Munaut <tnt@246tNt.com>
Date: Sun, 19 May 2024 14:33:08 +0000
Subject: [PATCH 7/8] defutil: Copy power pins placement status when applying
 DEF template

In newer OpenROAD, any power BTerm that isn't "FIXED" will be removed
completely if it's not connected when `pdngen` is called.

In OpenFrame those get connected later (in TinyTapeout case by some
custom script later and in the official example case using a macro).

So to work around that, we changed the DEF template to make sure all
power pins have the FIXED placement attribute and we make sure to copy
it when applying the template.

Signed-off-by: Sylvain Munaut <tnt@246tNt.com>
---
 openlane/scripts/odbpy/defutil.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/openlane/scripts/odbpy/defutil.py b/openlane/scripts/odbpy/defutil.py
index 9381293..b3dc16e 100644
--- a/openlane/scripts/odbpy/defutil.py
+++ b/openlane/scripts/odbpy/defutil.py
@@ -182,6 +182,7 @@ def relocate_pins(db, input_lefs, template_def, permissive, copy_def_power=False
                         box.yMin(),
                         box.xMax(),
                         box.yMax(),
+                        template_pin.getPlacementStatus(),
                     )
                 )
 
@@ -322,7 +323,7 @@ def relocate_pins(db, input_lefs, template_def, permissive, copy_def_power=False
                 template_bterm_location_tuple[3],
                 template_bterm_location_tuple[4],
             )
-            output_new_bpin.setPlacementStatus("PLACED")
+            output_new_bpin.setPlacementStatus(template_bterm_location_tuple[5])
 
     if grid_errors:
         print(
-- 
2.45.1

